wolfcrypt/src/aes.c

```diff
@@ -11340,6 +11340,311 @@ int wc_AesXtsDecrypt(XtsAes* xaes, byte* out, const byte* in, word32 sz,
     return ret;
 }
 
+/* Same process as wc_AesXtsEncryptSector except that the sector is incremented
+ * by the sectorSz every two blocks.
+ *
+ * xaes     AES keys to use for block encrypt/decrypt
+ * out      output buffer to hold cipher text
+ * in       input plain text buffer to encrypt
+ * sz       size of both out and in buffers
+ * sector   value to use for tweak
+ * sectorSz size of each sector
+ *
+ * returns 0 on success
+ */
+int wc_AesXtsEncryptConsecutiveSectors(XtsAes* xaes, byte* out, const byte* in,
+        word32 sz, word64 sector, word32 sectorSz)
+{
+    int ret = 0;
+    word32 blocks = (sz / AES_BLOCK_SIZE);
+    Aes *aes, *tweak;
+    byte* pt;
+    byte  i[AES_BLOCK_SIZE];
+
+    if (xaes == NULL || out == NULL || in == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    XMEMSET(i, 0, AES_BLOCK_SIZE);
+#ifdef BIG_ENDIAN_ORDER
+    sector = ByteReverseWord64(sector);
+#endif
+    pt = (byte*)&sector;
+    XMEMCPY(i, pt, sizeof(word64));
+
+    aes   = &xaes->aes;
+    tweak = &xaes->tweak;
+
+    if (blocks > 0) {
+        byte tmp[AES_BLOCK_SIZE];
+
+        XMEMSET(tmp, 0, AES_BLOCK_SIZE); /* set to 0's in case of improper AES
+                                          * key setup passed to encrypt direct*/
+
+        SAVE_VECTOR_REGISTERS(return _svr_ret;);
+
+        ret = wc_AesEncryptDirect(tweak, tmp, i);
+
+        if (ret != 0) {
+            RESTORE_VECTOR_REGISTERS();
+            return ret;
+        }
+
+        while (blocks > 0) {
+            word32 j;
+            byte carry = 0;
+
+            { /* check for if inline */
+                byte buf[AES_BLOCK_SIZE];
+
+                XMEMCPY(buf, in, AES_BLOCK_SIZE);
+                xorbuf(buf, tmp, AES_BLOCK_SIZE);
+                ret = wc_AesEncryptDirect(aes, out, buf);
+                if (ret != 0) {
+                    RESTORE_VECTOR_REGISTERS();
+                    return ret;
+                }
+            }
+            xorbuf(out, tmp, AES_BLOCK_SIZE);
+
+            /* multiply by shift left and propagate carry */
+            for (j = 0; j < AES_BLOCK_SIZE; j++) {
+                byte tmpC;
+
+                tmpC   = (tmp[j] >> 7) & 0x01;
+                tmp[j] = (byte)((tmp[j] << 1) + carry);
+                carry  = tmpC;
+            }
+            if (carry) {
+                tmp[0] ^= GF_XTS;
+            }
+
+            in  += AES_BLOCK_SIZE;
+            out += AES_BLOCK_SIZE;
+            sz  -= AES_BLOCK_SIZE;
+            blocks--;
+
+            if (blocks % (AES_BLOCK_SIZE * 2) == 0) {
+                XMEMSET(tmp, 0, AES_BLOCK_SIZE);
+                XMEMSET(i, 0, AES_BLOCK_SIZE);
+
+                sector += sectorSz;
+                pt = (byte*)&sector;
+                XMEMCPY(i, pt, sizeof(word64));
+
+                ret = wc_AesEncryptDirect(tweak, tmp, i);
+
+                if (ret != 0) {
+                    RESTORE_VECTOR_REGISTERS();
+                    return ret;
+                }
+
+            }
```

Comment:
This is the only addition over wc_AesXtsEncrypt right? If so lets reduce maintenance by having an internal XTS function that has a parameter for incrementing the sector value. 

For example:

```
static int wc_AesXTSEncryptInternal(XtsAes* xaes, byte* out, const byte* in, word32 sz,        
        const byte* i, word32 iSz, word32 sector, int sectorSz, int multSectors)
```

Where wc_AesXtsEncrypt would then call wc_AesXTSEncryptInternal with the last 3 arguments being 0 and this increment of sector value only happening if the multiSectors flag is used.